### PR TITLE
Add a new keyboard "Unicorne"

### PR DESCRIPTION
**Pull request by yanghu**
## Description
Adding a new custom keyboard to qmk. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by bcat** [comment]
FWIW, I've tested this on my Unicorne and it appears to work fine. :) Settings don't persist across reboots of the board, but I believe that's to be expected since QMK doesn't currently have an EEPROM emulation driver for the STM32F411. (I'm not knowledge enough about ARM to say if one could be developed or not... I know some of the other ARM chips do have working EEPROM emulation.)

**Comment by yanghu** [comment]
> FWIW, I've tested this on my Unicorne and it appears to work fine. :) Settings don't persist across reboots of the board, but I believe that's to be expected since QMK doesn't currently have an EEPROM emulation driver for the STM32F411. (I'm not knowledge enough about ARM to say if one could be developed or not... I know some of the other ARM chips do have working EEPROM emulation but I don't know the prerequisites.)

Yes, the keyboard is lacking eeprom, which was an oversight during design. However, I do have simple "patch" ideas: since the OLED headers/pins are i2c, it should be easy to design a small "eeprom" pcb, with only an external i2c eeprom chip. The small pcb can be soldered to the same pins, on the back side ( if you don't cut the pins of the oled header)

Personally I don't use eeprom. I do have a new board with an eeprom chip connected to i2c. once i have verified the circuit I can create the pcb described above. It should cost only a few bucks to order the pcb. (and the eeprom)

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/yanghu/unicorne/glcdfont.c [hunk 1]
```diff
@@ -0,0 +1,239 @@
+#pragma once
+
+#ifdef __AVR__
+ #include <avr/io.h>
+ #include <avr/pgmspace.h>
+#elif defined(ESP8266)
+ #include <pgmspace.h>
+#else
+ #define PROGMEM
+#endif
```
Comment (drashna) [inline keyboards/yanghu/unicorne/glcdfont.c:10]:
```suggestion
#include "progmem.h"
```

File: keyboards/yanghu/unicorne/glcdfont.c [hunk 2]
```diff
@@ -0,0 +1,239 @@
+#pragma once
```
Comment (drashna) [inline keyboards/yanghu/unicorne/glcdfont.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yanghu/unicorne/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            rgb_matrix_step_reverse();
+#else
+            rgblight_decrease_hue_noeeprom();
+#endif
+        }
+    }
+}
+
+// Set underglow color to blue.
+void keyboard_post_init_user(void) { rgblight_sethsv(HSV_BLUE); }
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/keymap.c:99]:
```suggestion
void keyboard_post_init_user(void) { rgblight_sethsv_noeeprom(HSV_BLUE); }
```

Also, you could use a define to set the default hue, iirc.

File: keyboards/yanghu/unicorne/keymaps/default/oled.c [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+void oled_task_user(void) {
+    /* oled_render_keylog(); */
+    oled_render_layer();
+    oled_render_mods();
+}
+
+// Call this from "led_update_user" and use `led_state.caps_lock` to check
+// the status of capslock.
+void oled_render_capslock(bool caps_on) {
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/oled.c:130]: This should be done in the oled_task_user function

You can use `host_keyboard_led_state()` to get the current status.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/yanghu/unicorne/keymaps/default/oled.c (hunk 4 above)
Comment (yanghu) [reply keyboards/yanghu/unicorne/keymaps/default/oled.c:124]: fixed in c4ea096

File: keyboards/yanghu/unicorne/keymaps/default/keymap.c (hunk 3 above)
Comment (yanghu) [reply keyboards/yanghu/unicorne/keymaps/default/keymap.c]: fixed in c4ea096

File: keyboards/yanghu/unicorne/glcdfont.c (hunk 2 above)
Comment (yanghu) [reply keyboards/yanghu/unicorne/glcdfont.c:16]: fixed in c4ea096

File: keyboards/yanghu/unicorne/glcdfont.c (hunk 1 above)
Comment (yanghu) [reply keyboards/yanghu/unicorne/glcdfont.c]: fixed in c4ea096

File: keyboards/yanghu/unicorne/unicorne.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+
+// clang-format off
+#define LAYOUT_unicorne( \
```
[the remark this answers is not in the document]
Comment (yanghu) [reply keyboards/yanghu/unicorne/unicorne.h]: fixed in c4ea096

File: keyboards/yanghu/unicorne/unicorne.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+) \
+LAYOUT_unicorne( \
+    KC_##K00, KC_##K01, KC_##K02, KC_##K03, KC_##K04, KC_##K05,      KC_##K06, KC_##K07, KC_##K08, KC_##K09, KC_##K0A, KC_##K0B, \
+    KC_##K10, KC_##K11, KC_##K12, KC_##K13, KC_##K14, KC_##K15,      KC_##K16, KC_##K17, KC_##K18, KC_##K19, KC_##K1A, KC_##K1B, \
+    KC_##K20, KC_##K21, KC_##K22, KC_##K23, KC_##K24, KC_##K25,      KC_##K26, KC_##K27, KC_##K28, KC_##K29, KC_##K2A, KC_##K2B, \
+           KC_##K30, KC_##K31, KC_##K32, KC_##K33,  KC_##K34, KC_##K35, KC_##K36, KC_##K37 \
+)
+// clang-format on
+
+#define LAYOUT LAYOUT_unicorne
```
[the remark this answers is not in the document]
Comment (yanghu) [reply keyboards/yanghu/unicorne/unicorne.h]: fixed in c4ea096

File: keyboards/yanghu/unicorne/readme.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+A compact 40% (3x6\_4) single-piece angled ortholinear ergo keyboard kit
+inspired by crkbd, Kyria, Ferris and Reviung41. Features OLED display, audio
+buzzer, RGB LED underglow as well as encoders.
+
+* Keyboard Maintainer: [Yang Hu](https://github.com/yanghu) 
+* Hardware Supported: Unicorne PCB 1.0 
+* Hardware Availability: Open source design at
+https://github.com/yanghu/unicorne
+
+Make example for this keyboard (after setting up your build environment):
```
[the remark this answers is not in the document]
Comment (yanghu) [reply keyboards/yanghu/unicorne/readme.md:18]: fixed in c4ea096

**Comment by yanghu** [comment]
I've fixed issues in c4ea096
Please take another look. Thank you!

File: keyboards/yanghu/unicorne/keymaps/default/rules.mk [hunk 8]
```diff
@@ -0,0 +1,3 @@
+ifeq ($(strip $(OLED_DRIVER_ENABLE)), yes)
+  SRC += oled.c
+endif
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/rules.mk:3]: This may not work well, would be simpler to always include it, and wrap the file in ifdefs.
```suggestion
SRC += oled.c
```

File: keyboards/yanghu/unicorne/keymaps/default/oled.c [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/oled.c:17]:
```suggestion
#ifdef OLED_DRIVER_ENABLE
#include QMK_KEYBOARD_H
```

File: keyboards/yanghu/unicorne/keymaps/default/oled.c [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    }
+}
+
+void oled_task_user(void) {
+    /* oled_render_keylog(); */
+    oled_render_layer();
+    oled_render_mods();
+    led_t led_state = host_keyboard_led_state();
+    oled_render_capslock(led_state.caps_lock);
+}
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/oled.c:141]:
```suggestion
}
#endif
```

**Review by drashna — COMMENTED** [review]
Also, it may be worth moving the oled code to `unicorne.c`, and set weakly.  That way, it will work on the configurator.

**Comment by yanghu** [comment]
Moved common oled code to `unicorne.c` and set then weakly so keymaps can override them.

**Comment by yanghu** [comment]
Changes were made on June 21 based on previous review. Please take a look at the new version, thanks!

File: keyboards/yanghu/unicorne/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    RGB_VAD,   RGB_SAD, RGB_HUD,  RGB_RMOD, XXXXXXX,   XXXXXXX,            XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,
+    XXXXXXX,   XXXXXXX, XXXXXXX,  XXXXXXX,  XXXXXXX,   XXXXXXX,            RESET,    XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,  XXXXXXX,
+             _______,  _______,   _______,   _______,  _______,   _______,   _______,  _______
+  ),
+};
+// clang-format on
+layer_state_t layer_state_set_user(layer_state_t state) { return update_tri_layer_state(state, _LOWER, _RAISE, _ADJUST); }
+
+// Left encoder scrolls the mousewheel. Right encoder adjusts underglow hue.
+void encoder_update_user(uint8_t index, bool clockwise) {
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/keymap.c:60]:
```suggestion
bool encoder_update_user(uint8_t index, bool clockwise) {
```

File: keyboards/yanghu/unicorne/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+#endif
+        } else {
+#ifdef RGB_MATRIX_ENABLE
+            rgb_matrix_step_reverse();
+#else
+            rgblight_decrease_hue_noeeprom();
+#endif
+        }
+    }
+}
```
Comment (drashna) [inline keyboards/yanghu/unicorne/keymaps/default/keymap.c:90]:
```suggestion
    return false;
}
```

File: keyboards/yanghu/unicorne/unicorne.c [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "unicorne.h"
+
+void matrix_init_kb(void) { matrix_init_user(); }
+
+void matrix_scan_kb(void) { matrix_scan_user(); }
```
Comment (drashna) [inline keyboards/yanghu/unicorne/unicorne.c:20]:
```suggestion
```

File: keyboards/yanghu/unicorne/readme.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+* Hardware Availability: Open source design at
+https://github.com/yanghu/unicorne
+
+To enter bootloader, long press the reset button for 3-5 seconds then release.
+More details can be found at [keyboard
+wiki](https://github.com/yanghu/unicorne/wiki/Build-guide#testing-the-board). 
+
+Make example for this keyboard (after setting up your build environment):
+
+		qmk compile -kb yanghu/unicorne -km default
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/readme.md:20]:
```suggestion
    make yanghu/unicorne:default
```

File: keyboards/yanghu/unicorne/readme.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+
+See the [build environment
+setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make
+instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more
+information.
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/readme.md:26]:
```suggestion
See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```

File: keyboards/yanghu/unicorne/readme.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+buzzer, RGB LED underglow as well as encoders.
+
+* Keyboard Maintainer: [Yang Hu](https://github.com/yanghu) 
+* Hardware Supported: Unicorne PCB 1.0 
+* Hardware Availability: Open source design at
+https://github.com/yanghu/unicorne
+
+To enter bootloader, long press the reset button for 3-5 seconds then release.
+More details can be found at [keyboard
+wiki](https://github.com/yanghu/unicorne/wiki/Build-guide#testing-the-board).
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/readme.md:16]:
```suggestion
A compact 40% (3x6\_4) single-piece angled ortholinear ergo keyboard kit inspired by crkbd, Kyria, Ferris and Reviung41. Features OLED display, audio buzzer, RGB LED underglow as well as encoders.

* Keyboard Maintainer: [Yang Hu](https://github.com/yanghu)
* Hardware Supported: Unicorne PCB 1.0
* Hardware Availability: Open source design at https://github.com/yanghu/unicorne

To enter bootloader, long press the reset button for 3-5 seconds then release. More details can be found at [keyboard wiki](https://github.com/yanghu/unicorne/wiki/Build-guide#testing-the-board). 
```

File: keyboards/yanghu/unicorne/readme.md [hunk 17]
```diff
@@ -0,0 +1,26 @@
+# Unicorne
+
+![Unicorne](https://imgur.com/g1nN3b4.jpg)
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/readme.md:3]:
```suggestion
![Unicorne](https://imgur.com/g1nN3b4h.jpg)
```

File: keyboards/yanghu/unicorne/keymaps/default/oled.h [hunk 18]
```diff
@@ -0,0 +1,17 @@
+/* Copyright 2021 Yang Hu
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/keymaps/default/oled.h:1]: This file can be deleted.

File: keyboards/yanghu/unicorne/keymaps/default/oled.c [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifdef OLED_DRIVER_ENABLE
+#    include QMK_KEYBOARD_H
+#    include "layers.h"
+#    include "oled.h"
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/keymaps/default/oled.c:19]:
```suggestion
```

File: keyboards/yanghu/unicorne/f411/rules.mk [hunk 20]
```diff
@@ -0,0 +1,7 @@
+# MCU name
+MCU = STM32F411
+
+# Address of the bootloader in system memory
+# * It is chip dependent, the correct number can be looked up here: (page 359)
+# * http://www.st.com/web/en/resource/technical/document/application_note/CD00167594.pdf
+STM32_BOOTLOADER_ADDRESS = 0x1FFF0000
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/f411/rules.mk:7]:
```suggestion

# Bootloader selection
BOOTLOADER = stm32-dfu
```

File: keyboards/yanghu/unicorne/rules.mk [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+WS2812_DRIVER = pwm
+AUDIO_DRIVER = pwm_hardware
+
+RGB_MATRIX_DRIVER = WS2812
+# Underglow and rgb matrix features shouldn't be on at the same time.
+# Choose one. Otherwise both driver will try to change color and you'll see
+# colors flickering.
+RGBLIGHT_ENABLE = yes # Enable keyboard RGB underglow
+RGB_MATRIX_ENABLE = no 
+
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/rules.mk:20]:
```suggestion
# Build Options
#   change yes to no to disable
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = yes          # Audio output
ENCODER_ENABLE = yes
OLED_DRIVER_ENABLE = yes

WS2812_DRIVER = pwm
AUDIO_DRIVER = pwm_hardware

RGB_MATRIX_ENABLE = no # Do not enable with RGBLIGHT
RGB_MATRIX_DRIVER = WS2812
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by yanghu — COMMENTED** [review]
@drashna FYI. 
Sorry about making changes after approval. I tried to verify the changes by building locally and found this issue. 

@fauxpark After commited this suggested change, I got the following build error. 

reverting this change would allow me to have a success build.

Shall I revert this change, or does it require sync to latest qmk:master first?

Thanks!

```
Linking: .build/yanghu_unicorne_f411_default.elf                                                    [ERRORS]
 |
 | /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: .build/obj_yanghu_unicorne_f411_default/protocol/chibios/main.o: in function `early_hardware_init_pre':
 | /home/huyang/repo/qmk_firmware/tmk_core/protocol/chibios/main.c:118: undefined reference to `enter_bootloader_mode_if_requested'
 | collect2: error: ld returned 1 exit status
 |
```

File: keyboards/yanghu/unicorne/f411/rules.mk [hunk 22]
```diff
@@ -1,7 +1,5 @@
 # MCU name
 MCU = STM32F411
 
-# Address of the bootloader in system memory
-# * It is chip dependent, the correct number can be looked up here: (page 359)
-# * http://www.st.com/web/en/resource/technical/document/application_note/CD00167594.pdf
-STM32_BOOTLOADER_ADDRESS = 0x1FFF0000
+# Bootloader selection
```
Comment (yanghu) [inline keyboards/yanghu/unicorne/f411/rules.mk:4]: @drashna FYI.

@fauxpark After commited this suggested change, I got the following build error. 

reverting this change would allow me to have a success build.

Shall I revert this change, or does it require sync to latest qmk:master first?

Thanks!

```
Linking: .build/yanghu_unicorne_f411_default.elf                                                    [ERRORS]
 |
 | /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: .build/obj_yanghu_unicorne_f411_default/protocol/chibios/main.o: in function `early_hardware_init_pre':
 | /home/huyang/repo/qmk_firmware/tmk_core/protocol/chibios/main.c:118: undefined reference to `enter_bootloader_mode_if_requested'
 | collect2: error: ld returned 1 exit status
 |
```

Comment (drashna) [reply keyboards/yanghu/unicorne/f411/rules.mk:4]: You may need to update your repo to get it to work properly.

Also, gcc 9 is prone to compiler issues.

Comment (bcat) [reply keyboards/yanghu/unicorne/f411/rules.mk:4]: I [rebased this PR onto QMK master](https://github.com/bcat/qmk_firmware/commits/unicorne) and it seems to build just fine. (I dropped the merged commits for a cleaner history, but I guess another `git merge master` would work just as well.) Haven't tested on a physical keyboard yet, but can do so tomorrow.

**Review by bcat — CHANGES_REQUESTED** [review]
Looks like the OLED configuration has bitrotted a bit. This seems to fix it, at least for the default keymap.

File: keyboards/yanghu/unicorne/keymaps/default/oled.c [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifdef OLED_DRIVER_ENABLE
```
Comment (bcat) [inline keyboards/yanghu/unicorne/keymaps/default/oled.c:16]:
```suggestion
#ifdef OLED_ENABLE
```

File: keyboards/yanghu/unicorne/rules.mk [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = yes          # Audio output
+ENCODER_ENABLE = yes
+OLED_DRIVER_ENABLE = yes
```
Comment (bcat) [inline keyboards/yanghu/unicorne/rules.mk:17]:
```suggestion
OLED_ENABLE = yes
OLED_DRIVER = SSD1306
```

File: keyboards/yanghu/unicorne/unicorne.c [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+  {208, 44},
+  {130, 60}
+}, {// LED Index to Flag
+  LED_FLAG_ALL, LED_FLAG_ALL, LED_FLAG_ALL, LED_FLAG_ALL, LED_FLAG_ALL, 
+  LED_FLAG_ALL, LED_FLAG_ALL, LED_FLAG_ALL
+}};
+// clang-format on
+#endif
+
+#ifdef OLED_DRIVER_ENABLE
```
Comment (bcat) [inline keyboards/yanghu/unicorne/unicorne.c:67]:
```suggestion
#ifdef OLED_ENABLE
```

File: keyboards/yanghu/unicorne/unicorne.h [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+#ifdef OLED_DRIVER_ENABLE
```
Comment (bcat) [inline keyboards/yanghu/unicorne/unicorne.h:21]:
```suggestion
#ifdef OLED_ENABLE
```

**Review by bcat — CHANGES_REQUESTED** [review]

File: keyboards/yanghu/unicorne/f411/rules.mk (hunk 22 above)
Comment (bcat) [reply keyboards/yanghu/unicorne/f411/rules.mk:4]: Okay, I've tested the latest version of this PR on my Unicorne. After rebasing against master and apply the `OLED_DRIVER` suggestions I left, everything works just fine on the physical keeb. :)

File: keyboards/yanghu/unicorne/info.json [hunk 27]
```diff
@@ -0,0 +1,64 @@
+{
+  "keyboard_name": "Unicorne",
+  "url": "https://github.com/yanghu/unicorne",
+  "maintainer": "yanghu",
+  "width": 15,
+  "height": 4.7,
```
Comment (fauxpark) [inline keyboards/yanghu/unicorne/info.json:6]:
```suggestion
```

**Comment by drashna** [comment]
@bcat Yeah, the F4x1 doesn't have EEPROM emulation currently.  IIRC, it's either on develop, or will be sooner or later.

**Review by tzarc — APPROVED** [review]
